src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -2036,13 +2036,26 @@ JvmtiEnvBase::force_early_return(jthread thread, jvalue value, TosState tos) {
   oop thread_obj = nullptr;
   jvmtiError err = get_threadOop_and_JavaThread(tlh.list(), thread, &java_thread, &thread_obj);
 
-  if (thread_obj != nullptr && thread_obj->is_a(vmClasses::BaseVirtualThread_klass())) {
-    // No support for virtual threads (yet).
-    return JVMTI_ERROR_OPAQUE_FRAME;
-  }
   if (err != JVMTI_ERROR_NONE) {
     return err;
   }
+  bool is_virtual = thread_obj != nullptr && thread_obj->is_a(vmClasses::BaseVirtualThread_klass());
```

Comment:
Does it make sense to reduce code duplication by moving these checks from forceearlyreturn and popframe code into a separate method?

Reply:
Good suggestion. Let me think a little bit. At least a merge with the PopFrame related push is needed first.